rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,100 @@
+---
+authors: Vladimir Kochnev (marshall-lee@evilmartians.com)
+state: draft
+---
+
+# RFD 1 - Kubernetes manager
+
+## What
+
+Teleport Kubernetes Manager is a [controller for Kubernetes](https://kubernetes.io/docs/concepts/architecture/controller/) serving Teleport resources defined as [Kubernetes custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+
+## Why
+
+Automated bootstrap of Teleport data is not an easy task, especially when it's running in Kubernetes. One can use `tctl` from the Teleport CLI toolset to bootstrap users/roles but integrate it into something like [Helm charts](https://helm.sh/docs/topics/charts/) quickly becomes a pain. In many scenarios, it's far more convenient to have the Teleport data defined the same way as the Kubernetes resources like Deployments and Services, in a format very similar to Teleport's own format.
+
+## Details
+
+### Implementation
+
+For the task, the [Kubebuilder](https://kubebuilder.io/) framework should be used. It has great code generation capabilities and gives a complete setup to develop & deploy the manager.
+
+### Instance definitions
+
+To write something into Teleport, the manager needs to know what server to connect to and how. So it needs Teleport Auth/Proxy/Tunnel address and credentials to be recognized. 
+
+Instance definitions look like this:
+
+```yaml
+apiVersion: instances.goteleport.com/v7
```

Comment:
> You might want to add a version field to the spec describing the Teleport version instead.

I thought about it — not in a context of `Instance` resources but in a context of `resources.goteleport.com` resources and it's problematic.

1. `apiVersion` + `version` creates a confusion for a user.
1. different `version` would probably mean different `spec` schema. CRDs doen't allow to make such a divergence — openapi v3 schema of a CRD is static cause it's used for validation etc

Reply:
Version here should be independent of Teleport version I think, otherwise we'd have to bump it for every release which is additional maintenance burden.

Reply:
Yeah probably we can make it independent. Frankly, I don't have a strong preference here. It just follows a scheme we use in all the `teleport-plugins` stuff. Initially it was proposed by @klizhentas that all the plugin versions should match the current Teleport version no matter what. So the choice to start `instance.goteleport.com` versioning from `v7` follows this scheme.